### DISCUSSION TITLE
Hosting outside of Docker and Azure Functions?

### DISCUSSION THREAD

**Discussion by louisfischer** [tfsaggregator/aggregator-cli#249] [Q&A]
Is there any guidance in running the aggregator or hosting it in IIS vs docker? My current employer does not support docker and all functionality has to be hosted in-house and not on azure.

**Comment by giuliov** [comment]
Guidance has to be written, maybe you can help by sharing your notes. Ideally a PR to [aggregator-docs](https://github.com/tfsaggregator/aggregator-docs).
I sketch here the process.

Basically, you compile the code yourself in Visual Studio or else. Build the server app `src/aggregator-host/aggregator-host.csproj`. That is a self-hosted version of the function using Kestrel. To run the app as a service you can use [nssm](https://nssm.cc/) or [winsw](https://github.com/winsw/winsw). You still need the PAT and SSL Certificate, prepare the `apikeys.json`
These steps are described in [Docker configuration](https://tfsaggregator.github.io/docs/v3/setup/docker) page. In addition to the listed [Environment variables](https://tfsaggregator.github.io/docs/v3/setup/docker/#environment-variables), set the environment variable `DOTNET_RUNNING_IN_CONTAINER` to `false` before launching the app.

Hope this helps you get started at least.

**Reply by louisfischer** [reply]
@giuliov  Thanks for guidance. I downloaded the code and tried to run the code manually. It seems that without modifying the Program code, the application  src/aggregator-host/aggregator-host.csproj will not work. Any thoughts on that?

![image](https://user-images.githubusercontent.com/1905010/147182097-82bdf1e5-92f4-482f-b315-0e722ff41627.png)

**Reply by giuliov** [reply]
No, I debug locally in Visual Studio by setting these environment variables

```
ASPNETCORE_ENVIRONMENT=Development
ASPNETCORE_Kestrel__Certificates__Default__Password=MY_CERTIFICATE_PASSWORD
ASPNETCORE_Kestrel__Certificates__Default__Path=PATH_TO_CERTIFICATE_IN_PFX_FORMAT
ASPNETCORE_URLS=http:////localhost:5321;https:////localhost:5320
Aggregator_ApiKeysPath=PATH_TO_APIKEYS.JSON
Aggregator_RulesPath=PATH_TO_RULES_DIRECTORY
Aggregator_SharedSecret=MY_SHARED_SECRET
Aggregator_VstsToken=MY_PERSONAL_ACCESS_TOKEN
Aggregator_VstsTokenType=PAT
DOTNET_RUNNING_IN_CONTAINER=true
Logging__LogLevel__Aggregator=Debug
```
